testing/async/mixin.h

```diff
@@ -0,0 +1,645 @@
+    iterator cbegin() const { return iterator{static_cast<value_type>(0)}; }
+
+    iterator end() { return iterator{static_cast<value_type>(num_values)}; }
+    iterator end() const { return iterator{static_cast<value_type>(num_values)}; }
+    iterator cend() const { return iterator{static_cast<value_type>(num_values)}; }
+
+    std::size_t size() const { return num_values; }
+  };
+
+  static input_type generate_input(std::size_t num_values)
+  {
+    return {num_values};
+  }
+};
+
+template <typename value_type>
+struct counting_iterator_1
+{
+  struct input_type
+  {
+    using iterator = thrust::counting_iterator<value_type>;
+
+    std::size_t num_values;
+
+    iterator begin() { return iterator{static_cast<value_type>(1)}; }
+    iterator begin() const { return iterator{static_cast<value_type>(1)}; }
+    iterator cbegin() const { return iterator{static_cast<value_type>(1)}; }
+
+    iterator end() { return iterator{static_cast<value_type>(1 + num_values)}; }
+    iterator end() const { return iterator{static_cast<value_type>(1 + num_values)}; }
+    iterator cend() const { return iterator{static_cast<value_type>(1 + num_values)}; }
+
+    std::size_t size() const { return num_values; }
+  };
+
+  static input_type generate_input(std::size_t num_values)
+  {
+    return {num_values};
+  }
+};
+
+template <typename value_type>
+struct constant_iterator_1
+{
+  struct input_type
+  {
+    using iterator = thrust::constant_iterator<value_type>;
+
+    std::size_t num_values;
+
+    iterator begin() { return iterator{static_cast<value_type>(1)}; }
+    iterator begin() const { return iterator{static_cast<value_type>(1)}; }
+    iterator cbegin() const { return iterator{static_cast<value_type>(1)}; }
+
+    iterator end() { return iterator{static_cast<value_type>(1)} + num_values; }
+    iterator end() const
+    {
+      return iterator{static_cast<value_type>(1)} + num_values;
+    }
+    iterator cend() const
+    {
+      return iterator{static_cast<value_type>(1)} + num_values;
+    }
+
+    std::size_t size() const { return num_values; }
+  };
+
+  static input_type generate_input(std::size_t num_values)
+  {
+    return {num_values};
+  }
+};
+
+} // namespace input
+
+//------------------------------------------------------------------------------
+namespace output
+{
+
+template <typename value_type>
+struct device_vector
+{
+  using output_type = thrust::device_vector<value_type>;
+
+  template <typename InputType>
+  static output_type generate_output(std::size_t num_values,
+                                     InputType& /* unused */)
+  {
+    return output_type(num_values);
+  }
+};
+
+template <typename value_type>
+struct device_vector_reuse_input
+{
+  using output_type = thrust::device_vector<value_type>&;
+
+  template <typename InputType>
+  static output_type generate_output(std::size_t /*num_values*/,
+                                     InputType& input)
+  {
+    return input;
+  }
+};
+
+struct discard_iterator
```

Comment:
@griwes suggests calling these `*_range`.

Reply:
The current name matches the name of what we're testing. I prefer this as-is.

It's scoped to a namespace so we can always differentiate this from the thrust class.